include/af/event.h

```diff
@@ -0,0 +1,111 @@
+/*******************************************************
+ * Copyright (c) 2019, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#pragma once
+
+#include <af/defines.h>
+
+#if AF_API_VERSION >= 37
+
+typedef void* af_event;
+
+#ifdef __cplusplus
+namespace af {
+
+/**
+    C++ RAII interface for manipulating events
+*/
+class AFAPI event {
+    af_event e_;
+
+   public:
+    event(af_event e);
+#if AF_COMPILER_CXX_RVALUE_REFERENCES
+    event(event&& other);
+    event& operator=(event&& other);
+#endif
+    event();
+    ~event();
+
+    af_event get() const;
+
+    void mark();
+
+    void enqueue();
+
+    void block();
+
+   private:
+    event& operator=(const event& other);
```

Comment:
Fair point. It's pretty unclear how the copy an event though since those semantics don't really exist for the underlying backend events (e.g. CUDA Event).